### PR TITLE
WIP: reduce unwanted type conversions of pd.Series.isin

### LINKED ISSUE #19356: .isin implicitly converts data types
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
pd.__version__
Out[1]: '0.22.0'

data = pd.Series([-9.0, 0.0])
data.isin([-9, -0.5])

Out[2]: 
0     True
1    False
dtype: bool

data = pd.Series([-9, 0])
data.isin([-9, -0.5])

Out[3]: 
0    True
1    True
dtype: bool


```
#### Problem description

.isin seems to implicitly convert the data type of what is passed in values depending on the data type of the Series. In the above example, I would expect the output to be (True, False) in both cases. There seems to have been a change in behavior after pandas version 0.19.2. With version 0.19.2 the output is in both cases (True, False).

#### Expected Output

```python
import pandas as pd
pd.__version__
Out[1]: '0.19.2'

data = pd.Series([-9.0, 0.0])
data.isin([-9, -0.5])

Out[2]: 
0     True
1    False
dtype: bool

data = pd.Series([-9, 0])
data.isin([-9, -0.5])

Out[3]: 
0    True
1    False
dtype: bool


```

### PR DESCRIPTION
**Pull request by jdoepfert**
Addresses the bug in pd.Series.isin(values) that sometimes undesirably converts the dtype of `values` to the the dtype of the series before the `isin` comparison. Here is an example where `values` is converted to `int`, and hence the output `(True, True)` is yielded, while `(True, False)` is expected:
```python
>>>  values = [1, 0.5]
>>> pd.Series([1, 0]).isin(values)
0    True
1    True
dtype: bool
```

In my understanding, the bug has two sources:

**1) L407-408 in `pandas/core/algorithms.py`**

```python
comps, dtype, _ = _ensure_data(comps)
values, _, _ = _ensure_data(values, dtype=dtype)
```
Here, first the dtype of the series (`comps`) is determined, and then `values` is forcefully converted to that dtype. This e.g. leads to the undesired conversion of the example above.

**2) L417-434 in `pandas/core/algorithms.py`**
```python
    elif is_integer_dtype(comps):
        try:
            values = values.astype('int64', copy=False)
            comps = comps.astype('int64', copy=False)
            f = lambda x, y: htable.ismember_int64(x, y)
        except (TypeError, ValueError):
            values = values.astype(object)
            comps = comps.astype(object)

    elif is_float_dtype(comps):
        try:
            values = values.astype('float64', copy=False)
            comps = comps.astype('float64', copy=False)
            checknull = isna(values).any()
            f = lambda x, y: htable.ismember_float64(x, y, checknull)
        except (TypeError, ValueError):
            values = values.astype(object)
            comps = comps.astype(object)
```
Here, if the dtype of the series (`comps`) is either `float` or `int`, a forced conversion of `values` takes place again.

The fix:
**1) L408ff**
Only force the dtype of `comps` to `values` if *both* are int-like, float-like, or datetime-like.  If not, use `_ensure_data(values)`  *without* the `dtype` argument. I tried to keep the cost of the dtype checks low by exploiting lazy evaluation.

**2) L432ff** 
Add `and` statements such that the conversion only takes place when *both* `comps` and `values` are either `float` or `int`. If they are none of the above is true, both are converted to `object` type in the `else` block. This ensures that `return f(comps, values)` works (since `f = lambda x, y: htable.ismember_object(x, values)` from L426).

- [x] fixes #19356
- [x] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/algorithms.py [hunk 1]
```diff
@@ -405,7 +406,21 @@ def isin(comps, values):
         values = construct_1d_object_array_from_listlike(list(values))
 
     comps, dtype, _ = _ensure_data(comps)
-    values, _, _ = _ensure_data(values, dtype=dtype)
+    # Convert `values` to `dtype` if dtype of `values` is like `dtype`
```
Comment (jreback) [inline pandas/core/algorithms.py]: you don't need to check all these conditions, rather you can just do

In [14]: l = [1.0, 1]

```
In [15]: set([type(v) for v in l])
Out[15]: {float, int}
```
If you have a set of len(1) then you can go ahead and pass the dtype (else you can use ``object``).

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a release note.

this absolutely needs a performance check. needs to run the asvs for algorithms

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jdoepfert) [reply pandas/core/algorithms.py]: @jreback thx for the comment, not sure if I got you right though. Imagine I have
```python
pd.Series([1, 0]).isin([1.0, 0.5])
```
Then the len of the set would be one (I only have floats), but the conversion to `int` would still not be desirable.

**Comment by jdoepfert** [comment]
@jreback Added a description. Agree, this definitely needs a performance check, will work on that tomorrow

File: pandas/core/algorithms.py (hunk 1 above)
Comment (jreback) [reply pandas/core/algorithms.py]: thats a different case which is actually simple. the dtype that ensure_data returns is different from the dtype that the set check returns. 

NONE of this checking can be done here, rather it is *already* done in ensure_data

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jdoepfert** [comment]
Still not sure I got it. Tried to implement the set check approach, but due to this test (L502 in `test_algos.py`)
```python
    def test_i8(self):

        arr = pd.date_range('20130101', periods=3).values
        result = algos.isin(arr, [arr[0]])
        expected = np.array([True, False, False])
```
I still had to to some datetime-like type checking:
Here, the dtype of `comps` is initially `datetime64[ns]`, but ensure_data makes it `int64` (`<M8[ns]`).  The dtype of the items in `values` is different, so no conversion is enforced. 
Hence I do not only check for exact equality of the dtypes (that ensure_data and the set check return) but I additionally use `is_time_like()`. 

Some tests are still failing, but if that's the way to go, I'll fix this up

**Comment by jdoepfert** [comment]
Won't have time to work on this for the next 2 months, so if anyone wants to take over, give it a go! The last commit with all tests passing was https://github.com/pandas-dev/pandas/pull/19508/commits/cd14c56f5b3f8415e36b20934e8a011a462603d4.

**Comment by jreback** [comment]
closing as stale.

**Pull request closed without merging**